File: src/cpp/ee/services/internal/GenericAd.cpp
//
//  GenericAd.cpp
//  Pods
//
//  Created by eps on 6/2/21.
//

#include "ee/services/internal/GenericAd.hpp"

#include <ee/ads/internal/ICapper.hpp>
#include <ee/ads/internal/IRetrier.hpp>
#include <ee/core/NoAwait.hpp>
#include <ee/core/ObserverHandle.hpp>
#include <ee/core/Task.hpp>

namespace ee {
namespace services {
using Self = GenericAd;

Self::GenericAd(const std::shared_ptr<IAd>& ad,
                const std::shared_ptr<ads::ICapper>& loadCapper,
                const std::shared_ptr<ads::IRetrier>& loadRetrier)
    : ad_(ad)
    , loadCapper_(loadCapper)
    , loadRetrier_(loadRetrier) {
    handle_ = std::make_unique<ObserverHandle>();
    (*handle_).bind(*ad_).addObserver({
        .onLoaded =
            [this] {
                // Propagation.
                dispatchEvent([](auto&& observer) {
                    if (observer.onLoaded) {
                        observer.onLoaded();
                    }
                });
            },
        .onLoadResult =
            [this](const AdLoadResult& result) {
                // Propagation.
                dispatchEvent([result](auto&& observer) {
                    if (observer.onLoadResult) {
                        observer.onLoadResult(result);
                    }
                });
            },
        .onClicked =
            [this] {
                // Propagation.
                dispatchEvent([](auto&& observer) {
                    if (observer.onClicked) {
                        observer.onClicked();
                    }
                });
            },
    });
}

Self::~GenericAd() = default;

void Self::destroy() {
    ad_->destroy();
    handle_->clear();
    loadRetrier_->stop();
}

bool Self::isLoaded() const {
    return ad_->isLoaded();
}

Task<bool> Self::load() {
    auto result = co_await loadInternal();
    if (result) {
        loadRetrier_->stop();
    } else {
        noAwait([this]() -> Task<> {
            co_await loadRetrier_->process([this]() -> Task<bool> { //
                co_return co_await loadInternal();
            });
        });
    }
    co_return result;
}

Task<bool> Self::loadInternal() {
    if (loadCapper_->isCapped()) {
        co_return false;
    }
    loadCapper_->cap();
    co_return co_await ad_->load();
}
} // namespace services
} // namespace ee

File: src/unity/Runtime/Services/Internal/GenericAd.cs
using System.Threading.Tasks;

namespace EE.Internal {
    internal class GenericAd : ObserverManager<AdObserver>, IAd {
        private readonly IAd _ad;
        private readonly ICapper _loadCapper;
        private readonly IRetrier _loadRetrier;
        private readonly ObserverHandle _handle;

        protected GenericAd(
            IAd ad,
            ICapper loadCapper,
            IRetrier loadRetrier) {
            _ad = ad;
            _loadCapper = loadCapper;
            _loadRetrier = loadRetrier;
            _handle = new ObserverHandle();
            _handle.Bind(_ad)
                .AddObserver(new AdObserver {
                    OnLoaded = () => DispatchEvent(observer =>
                        observer.OnLoaded?.Invoke()),

                    OnLoadResult = result => DispatchEvent(observer =>
                        observer.OnLoadResult?.Invoke(result)),

                    OnClicked = () => DispatchEvent(observer =>
                        observer.OnClicked?.Invoke())
                });
        }

        public void Destroy() {
            _ad.Destroy();
            _handle.Clear();
            _loadRetrier.Stop();
        }

        public bool IsLoaded => _ad.IsLoaded;

        public async Task<bool> Load() {
            var result = await LoadInternal();
            if (result) {
                _loadRetrier.Stop();
            } else {
                Utils.NoAwait(async () => { //
                    await _loadRetrier.Process(async () => await LoadInternal());
                });
            }
            return result;
        }

        private async Task<bool> LoadInternal() {
            if (_loadCapper.IsCapped) {
                return false;
            }
            _loadCapper.Cap();
            return await _ad.Load();
        }
    }
}